### PR TITLE
Validate gasLimitBoundDivisor != 0

### LINKED ISSUE #7482: Parity crashes when attempting to start a private network
> I'm running:
>
> - **Which Parity version?**: 1.8.5-beta
> - **Which operating system?**: Windows / MacOS / __Linux__
> - **How installed?**: via installer / homebrew / __binaries__ / from source
> - **Are you fully synchronized?**: no / __yes___
> - **Did you try to restart the node?**: no / __yes__

I'm trying to run a private network using parity. I configured my genesis file to look like this:
```
{
    "name": "Bowery",
    "engine": {
        "authorityRound": {
            "params": {
                "stepDuration": "12",
                "validators": {
                    "list": [
                        "0x50b46ac692821cf0605192e452e28c8389e138ca",
                        "0x289386c7ebb6c5bc062cfc957e124d54a9aa1609"
                    ]
                },
                "blockReward": "8080808080808080808"
            }
        }
    },
    "genesis": {
        "seal": {
            "authorityRound": {
                "step": "0x0",
                "signature": "0x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
            }
        },
        "difficulty": "0x0",
        "gasLimit": "0x2fefd8"
    },
    "params": {
        "networkID": "0x67",
        "maximumExtraDataSize": "0x20",
        "minGasLimit": "0x1388",
        "gasLimitBoundDivisor": "0x0"
    },
    "accounts": {
        "0x0000000000000000000000000000000000000001": { "balance": "1", "builtin": { "name":   "ecrecover", "pricing": { "linear": { "base": 3000, "word": 0 } } } },
        "0x0000000000000000000000000000000000000002": { "balance": "1", "builtin": { "name": "sha256", "pricing": { "linear": { "base": 60, "word": 12 } } } },
        "0x0000000000000000000000000000000000000003": { "balance": "1", "builtin": { "name": "ripemd160", "pricing": { "linear": { "base": 600, "word": 120 } } } },
        "0x0000000000000000000000000000000000000004": { "balance": "1", "builtin": { "name": "identity", "pricing": { "linear": { "base": 15, "word": 3 } } } },
        "0x50b46ac692821cf0605192e452e28c8389e138ca": {
            "balance": "1000000000000000000"
        },
        "0x289386c7ebb6c5bc062cfc957e124d54a9aa1609": {
            "balance": "1000000000000000000"
        }
    }
}
```

I generated these validator accounts by fist inputting null addresses and running `parity --chain bowery_gen.json account new`. `bower_gen.json` is the relative path to the json file I pasted above. When starting the parity node with the command `parity --chain bower_gen.json` I receive the following message: 

```

stack backtrace:
   0:     0x560a2676ba1c - <no info>
Thread 'IO Worker #3' panicked at 'assertion failed: your_bits != 0', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/bigint-4.2.0/src/uint.rs:1227
This is a bug. Please report it at:
    https://github.com/paritytech/parity/issues/new
[1]    581 abort (core dumped)  parity --chain bowery_gen.json

```

I was following the documentation [here](https://paritytech.github.io/wiki/Chain-specification). The docs were a bit sparse so I may be doing something wrong here. Any guidance is appreciated!

### PR DESCRIPTION
**Pull request by asymmetric**
This PR adds a few things:

* validation against the `gasLimitBoundDivisor` being 0.
* a `spec::Error` type that wraps both parsing errors and semantic ones. 
* a couple of tests, one for the above case, another one for a missing field in the Spec.

Closes #7482.

### CODE REVIEW

File: json/src/spec/spec.rs [hunk 1]
```diff
@@ -44,14 +44,28 @@ pub struct Spec {
...
 	pub fn load<R>(reader: R) -> Result<Self, Error> where R: Read {
-		serde_json::from_reader(reader)
+		let spec: Self = serde_json::from_reader(reader)?;
+
+		Self::validate(&spec)?;
+
+		Ok(spec)
+	}
+
+	fn validate(spec: &Self) -> Result<(), Error> {
```
Comment (rphmeier) [inline json/src/spec/spec.rs]: `fn validate(&self)`

File: json/src/spec/spec.rs [hunk 2]
```diff
@@ -44,14 +44,28 @@ pub struct Spec {
 impl Spec {
 	/// Loads test from json.
 	pub fn load<R>(reader: R) -> Result<Self, Error> where R: Read {
-		serde_json::from_reader(reader)
+		let spec: Self = serde_json::from_reader(reader)?;
+
+		Self::validate(&spec)?;
```
Comment (rphmeier) [inline json/src/spec/spec.rs]: and then `spec.validate()?`

**Comment by andresilva** [comment]
I think that's a good idea, annotating every divisor with something like `#[serde(deserialize_with = "validate_non_zero")]` is clear enough for me. WDYT @rphmeier?

**Comment by rphmeier** [comment]
@andresilva sounds like a good approach to me, definitely an improvement over the current status.

**Comment by asymmetric** [comment]
I think I have some spare cycles this week to take care of this, unless there's already someone working on it.

**Comment by rphmeier** [comment]
@asymmetric feel free to take it. If you want, I can merge this in the meantime and the follow-up PR will do either the `NonZeroUint` or `deserialize_with` improvement.

**Pull request closed without merging**